File: challenge3/learning_codebase.js
class InvoiceProcessor {
  constructor(dbConn) {
    this.db = dbConn;
  }

  generateInvoice(userId, items) {
    const total = items.reduce((sum, item) => sum + item.price * item.quantity, 0);
    const invoiceId = this.db.insertInvoice(userId, total);
    items.forEach(item => {
      this.db.insertLineItem(invoiceId, item.name, item.price, item.quantity);
    });
    this.db.logInvoiceCreation(invoiceId);
    return invoiceId;
  }

  refundInvoice(invoiceId) {
    const lineItems = this.db.getLineItems(invoiceId);
    lineItems.forEach(item => {
      this.db.processRefund(item.name, item.price, item.quantity);
    });
    this.db.markInvoiceRefunded(invoiceId);
  }

  summarizeInvoice(invoiceId) {
    const items = this.db.getLineItems(invoiceId);
    const summary = {};
    items.forEach(item => {
      summary[item.name] = item.quantity;
    });
    return summary;
  }
}

File: challenge3/learning_codebase.py
# This is a file to be used for prompt engineering challenge 3

class InvoiceProcessor:
    def __init__(self, db_conn):
        self.db = db_conn

    def generate_invoice(self, user_id, items):
        total = sum(item['price'] * item['quantity'] for item in items)
        invoice_id = self.db.insert_invoice(user_id, total)
        for item in items:
            self.db.insert_line_item(invoice_id, item['name'], item['price'], item['quantity'])
        self.db.log_invoice_creation(invoice_id)
        return invoice_id

    def refund_invoice(self, invoice_id):
        line_items = self.db.get_line_items(invoice_id)
        for item in line_items:
            self.db.process_refund(item['name'], item['price'], item['quantity'])
        self.db.mark_invoice_refunded(invoice_id)

    def summarize_invoice(self, invoice_id):
        items = self.db.get_line_items(invoice_id)
        summary = {item['name']: item['quantity'] for item in items}
        return summary

